packages/react-stylesheets/src/StylesheetContext.tsx

```diff
@@ -0,0 +1,66 @@
+import * as React from 'react';
+export type HTMLDirection = 'rtl' | 'ltr' | 'auto';
+
+export interface StylesheetContextType {
+  dir: HTMLDirection;
+  target: Document | undefined;
+  register: (stylesheets: string[], context: StylesheetContextType) => void;
+  styleCache: WeakMap<Document, Map<string, Map<string, boolean>>>;
+  enqueuedSheets: string[];
+  renderSheets: (stylesheets: string[], context: StylesheetContextType) => void;
+  // hasRegistered: (stylehseet: string, context: StylesheetContextType) => boolean;
+}
+
+// Default registration function for stylesheets.
+export const register = (stylesheets: string[], context: StylesheetContextType) => {
+  const { styleCache, target, dir } = context;
+
+  // Grab the style cache for the target document.
+  let targetDirections = styleCache.get(target!);
+
+  if (!targetDirections) {
+    targetDirections = new Map();
+    styleCache.set(target!, targetDirections);
+  }
+
+  // Grab the sheet set for the correct direction.
+  let sheets = targetDirections.get(dir);
+
+  if (!sheets) {
+    sheets = new Map();
+    targetDirections.set(dir, sheets);
+  }
+
+  for (const stylesheet of stylesheets) {
+    if (!sheets.has(stylesheet)) {
+      context.enqueuedSheets.unshift(stylesheet);
+      sheets.set(stylesheet, true);
+    }
+  }
+
+  // If there is no target, call renderSheets immediately.
+  if (!target) {
+    renderSheets(context.enqueuedSheets, context);
+  }
```

Comment:
With server rendering, `renderSheets` will never get called if the primary entrypoint is through `useLayoutEffect`.

How would you expect the syntax to work for SSR scenarios? Here's an alternative `server` boolean we could apply, rather than a null `target` value.

```tsx
<StylesheetProvider server renderStyles={...}>...</StylesheetProvider>
  ...

Reply:
hm ic. maybe something like:
```
  const isSSR = !target;
  if (isSSR) {
    renderSheets(context.enqueuedSheets, context);
  }
```
instead of
```
  // If there is no target, call renderSheets immediately.
  if (!target) {
    renderSheets(context.enqueuedSheets, context);
  }
```

Reply:
makes sense!